framework/src/tinyalsa/tinyalsa.c

```diff
@@ -1109,53 +1188,134 @@ int pcm_stop(struct pcm *pcm)
 		return -EINVAL;
 	}
 #ifdef CONFIG_AUDIO_MULTI_SESSION
-	if (ioctl(pcm->fd, AUDIOIOC_STOP, (unsigned long)pcm->session) < 0)
+	if (!pcm->running && ioctl(pcm->fd, AUDIOIOC_STOP, (unsigned long)pcm->session) < 0)
 #else
-	if (ioctl(pcm->fd, AUDIOIOC_STOP, 0) < 0)
+	if (!pcm->running && ioctl(pcm->fd, AUDIOIOC_STOP, 0) < 0)
 #endif
 		return oops(pcm, errno, "cannot stop channel");
 
+	/* Remove any pending messages from the message queue */
+	struct timespec st_time;
+	struct audio_msg_s msg;
+	unsigned int size;
+	int prio;
+	while (1) {
+		clock_gettime(CLOCK_REALTIME, &st_time);
+		size = mq_timedreceive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio, &st_time);
+		if (size != sizeof(msg)) {
+			break;
+		}
+	}
+
 	pcm->prepared = 0;
 	pcm->running = 0;
-	pcm->bufPtr = 0;
-	pcm->nextSize = 0;
-	pcm->nextOffset = 0;
-	pcm->nextBuf = NULL;
+	pcm->draining = 0;
+	pcm->buf_ptr = 0;
+	pcm->next_size = 0;
+	pcm->next_offset = 0;
+	pcm->next_buf = NULL;
 
 	return 0;
 }
 
+/** Stops a PCM. Any data present in the buffers will be dropped.
+ * @param pcm A PCM handle.
+ * @return On success, zero; on failure, a negative number.
+ * @ingroup libtinyalsa-pcm
+ */
+int pcm_drop(struct pcm *pcm)
```

Comment:
I think drop is not same as stop
it just remove all data from buffer,
In this case if user call drop to clean buffer, they should start pcm again.
below is generic usage of pcm_drop, it calling pcm_close after drop.
https://www.alsa-project.org/alsa-doc/alsa-lib/_2test_2latency_8c-example.html

Reply:
Yes, you are correct. 
If user calls pcm_drop() he has to start the pcm again to send new data. 
I think the same thing is shown in the example of above link. 
Every time the while(1) loop is entered, they are calling snd_pcm_start(chandle).
Since they use snd_pcm_link(chandle, phandle), the single call to snd_pcm_start(chandle) will start both the pcm devices.

Also, according to the API reference, pcm_drop is same as pcm_stop.
Stop a PCM dropping pending frames.
This function stops the PCM immediately. The pending samples on the buffer are ignored.

Reply:
ok I will accept your opinion.